src/ocnicepost.fd/utils_mod.F90

```diff
@@ -0,0 +1,604 @@
+    rc = nf90_inq_dimid(ncid, 'n_s', id)
+    rc = nf90_inquire_dimension(ncid, id, len=n_s)
+    rc = nf90_inq_dimid(ncid, 'n_a', id)
+    rc = nf90_inquire_dimension(ncid, id, len=n_a)
+    rc = nf90_inq_dimid(ncid, 'n_b', id)
+    rc = nf90_inquire_dimension(ncid, id, len=n_b)
+
+    allocate(col(1:n_s)); col = 0
+    allocate(row(1:n_s)); row = 0
+    allocate(  S(1:n_s)); S = 0.0
+
+    rc = nf90_inq_varid(ncid, 'col', id)
+    rc = nf90_get_var(ncid,     id, col)
+    rc = nf90_inq_varid(ncid, 'row', id)
+    rc = nf90_get_var(ncid,     id, row)
+    rc = nf90_inq_varid(ncid,   'S', id)
+    rc = nf90_get_var(ncid,      id,  S)
+    rc = nf90_close(ncid)
+
+    dst_field = 0.0
+    do i = 1,n_s
+       ii = row(i); jj = col(i)
+       dst_field(ii) = dst_field(ii) + S(i)*real(src_field(jj),8)
+    enddo
+
+    if (debug) write(logunit,'(a)')'exit '//trim(subname)
+  end subroutine remap1d
+
+  !----------------------------------------------------------
+  ! remap a packed field of either nflds or nlevs
+  !----------------------------------------------------------
+  subroutine remap2d(fname, dim2, src_field, dst_field)
+
+    character(len=*), intent(in)  :: fname
+    integer,          intent(in)  :: dim2
+    real,             intent(in)  :: src_field(:,:)
+    real,             intent(out) :: dst_field(:,:)
+
+    ! local variables
+    integer :: ncid, rc, id
+    integer :: i,ii,jj
+    integer :: n_a, n_b, n_s
+    integer(kind=4), allocatable, dimension(:) :: col, row
+    real(kind=8),    allocatable, dimension(:) :: S
+    character(len=20) :: subname = 'remap2d'
+
+    if (debug)write(logunit,'(a)')'enter '//trim(subname)//' weights = '//trim(fname)
+
+    ! retrieve the weights
+    rc = nf90_open(trim(fname), nf90_nowrite, ncid)
+    call handle_err(rc,' nf90_open '//fname)
+    rc = nf90_inq_dimid(ncid, 'n_s', id)
+    rc = nf90_inquire_dimension(ncid, id, len=n_s)
+    rc = nf90_inq_dimid(ncid, 'n_a', id)
+    rc = nf90_inquire_dimension(ncid, id, len=n_a)
+    rc = nf90_inq_dimid(ncid, 'n_b', id)
+    rc = nf90_inquire_dimension(ncid, id, len=n_b)
+
+    allocate(col(1:n_s)); col = 0
+    allocate(row(1:n_s)); row = 0
+    allocate(  S(1:n_s)); S = 0.0
+
+    rc = nf90_inq_varid(ncid, 'col', id)
+    rc = nf90_get_var(ncid,     id, col)
+    rc = nf90_inq_varid(ncid, 'row', id)
+    rc = nf90_get_var(ncid,     id, row)
+    rc = nf90_inq_varid(ncid,   'S', id)
+    rc = nf90_get_var(ncid,      id,  S)
+    rc = nf90_close(ncid)
+
+    dst_field = 0.0
+    do i = 1,n_s
+       ii = row(i); jj = col(i)
+       dst_field(ii,:) = dst_field(ii,:) + S(i)*real(src_field(jj,:),8)
+    enddo
+
+    if (debug) write(logunit,'(a)')'exit '//trim(subname)
+  end subroutine remap2d
+
+  !----------------------------------------------------------
+  ! remap a field packed array of nk levels and nflds fields
+  !----------------------------------------------------------
+  subroutine remap3d(fname, nk, nflds, src_field, dst_field)
```

Comment:
There should not be separate subroutines for 2-D versus 3-D arrays. This should be accomplished using one subroutine which determines the z-coordinate dimension and then loops over the respective dimension to perform the remapping for the respective levels.